Initial support for upstream HTTP/1.1 tunneling

irozzo-1A opened the pull request:
Signed-off-by: Iacopo Rozzo <iacopo@kubermatic.com>

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

-->
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:
Additional Description:
Risk Level: Low
Testing: unit test, integration, manual testing
Docs Changes: Added documentation on how to configure Envoy for tunneling TCP over HTTP/1
Release Notes: n/a (still hidden)
Fixes #11308

irozzo-1A commented:
This is still WIP I have the following points to address:

- [ ] Fix suspect fault occurring when upstream do not respond with 200
- [ ] Make sure that HttpUpstream::isValidBytestreamResponse accepts well formed http connect responses as definded in [RFC7231](https://tools.ietf.org/html/rfc7231#section-4.3.6)
- [ ] Make sure that data is streamed to the destination only after 2xx message is received from upstream
- [ ] Add unit and integration tests
- [ ] Update docs
- [ ] (optional) Validate hostname provided with tunneling configuration has [authority form](https://tools.ietf.org/html/rfc7230#section-5.3)

alyssawilk commented:
Sweet, so excited to see progress on this!  lmk when you want me to take a pass!

irozzo-1A commented:
> Sweet, so excited to see progress on this! lmk when you want me to take a pass!

Thx @alyssawilk! I'll let you know. I think I'll have time to continue on this tomorrow or Friday on this. 

I have just a question. For HTTP2 you ensured that no data is sent before the response to the CONNECT is received by the upstream https://github.com/envoyproxy/envoy/pull/10974/. From what I observed the HTTP1 flow is different and the same logic does not work. Do you confirm that? Do you have an idea where I could implement this logic?

alyssawilk commented:
I think this is less about HTTP/2 and HTTP/1 and more about Enovy proxying CONNECT vs Envoy originating CONNECT.  Either way I think it'd be great to fix.

I think to do that we could defer onPoolReadyBase from Filter::onPoolReady and instead call it when HttpUpstream::isValidBytestreamResponse is confirmed.  That could be done as a standalone PR or with this one - whatever you prefer!

irozzo-1A commented:
Ok, you're right. I did not look carefully enough into this. I will leave it for another PR in this case as it is not exclusively related to HTTP/1.1 tunneling.

irozzo-1A commented:
Hey @alyssawilk, I found the time to progress a bit on this and I have two issues related to the integration tests I added so far. I would like to have your opinion on those:

- The first test is based on the [Basic](https://github.com/envoyproxy/envoy/blob/22683a0a24ffbb0cdeb4111eec5ec90246bec9cb/test/integration/tcp_tunneling_integration_test.cc#L283) TcpTunnelingIntegrationTest and is failing on the [waitForEndStream](https://github.com/envoyproxy/envoy/pull/13293/files#diff-0880b68fbe151043a6ba111cadd60370R603) assertion. This looks like a real issue, if my understanding is correct after the downstream connection is closed from the client side, and all data is transmitted to the upstream, the TCP connection with the upstream should be closed as well. From the test failure and the traffic capture I made it does not seem to be the case. Do you agree this is the expected behavior? Do you have a rough idea on how where this issue should be fixed?
- The second is based on [InvalidResponseHeaders](https://github.com/envoyproxy/envoy/blob/22683a0a24ffbb0cdeb4111eec5ec90246bec9cb/test/integration/tcp_tunneling_integration_test.cc#L331) TcpTunnelingIntegrationTest, it is failing on [waitForReset](https://github.com/envoyproxy/envoy/pull/13293/files#diff-0880b68fbe151043a6ba111cadd60370R623) assertion. In this case, it seems just to be an issue in the test itself. When the upstream respond with a 500 both connections with upstream and downstream are closed as expected. From what I observed so far the test failure seems to be related to [onResetStream]() method not being called on the [FakeStream](https://github.com/envoyproxy/envoy/blob/b50af650301e90eab60d14fa438a018edde31f0a/test/integration/fake_upstream.cc#L131), but I did not figure out the root cause yet.

alyssawilk commented:
Hm, the end stream issue is an interesting one.
In HTTP/2 there's a clear way to end the stream for non-chunked data, so if a connection is half closed on TCP (which we support), we can end-stream for HTTP/2, then further upstream if we downgrade back to TCP we can transfer that end-stream back into a half close.
For HTTP/1 there's equivalent way to end stream.  We could half-close the TCP connection, but Envoy specifically doesn't support half close for HTTP, so it'd result in a connection close at the far side rather than a reset stream.  I think the best thing we can do here is send a half close, and note in docs that CONNECT over HTTP/1.1 uses half-close semantics which will not fully work with servers which don't support half close over HTTP/1.1
(please snag me on slack if that doesn't make sense - I'm not  sure how familiar you are with all the above and I can go into as much detail as you need)

for waitForReset you may need to wait for a connection close with HTTP/1.1 since that's how stream resets are communicated.

alyssawilk commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -466,19 +466,19 @@ bool Filter::maybeTunnel(const std::string& cluster_name) {
-      upstream_ = std::make_unique<HttpUpstream>(*upstream_callbacks_,
-                                                 config_->tunnelingConfig()->hostname());
+      if ((cluster->info()->features() & Upstream::ClusterInfo::Features::HTTP2) == 0) {
+        // TODO(irozzo): How to detect HTTP3
```
I think you can replace the old early return-false with features() & HTTP3 != 0

irozzo-1A replied:
Unfortunately, as we discussed offline the only feature flag I see at the moment is about HTTP2: https://github.com/envoyproxy/envoy/blob/master/include/envoy/upstream/upstream.h#L718.
Anyway, HTTP3 upstreams do not seem to be supported yet.

alyssawilk reviewed (commented):
looking better and better. :-)
one drive by comment but otherwise I'm going to wait for you to tell me I should do another pass.  Just ping me when you're ready!

irozzo-1A commented:
I'm summarizing here the issues are blocking me at the moment.

The first issue I met is that, the downstream client closing the connection (i.e. sending FIN) was not resulting in the downstream connection being closed ([this](https://github.com/envoyproxy/envoy/blob/05e3529ebab0035310e8fd284ea5f05c7f3d27d8/test/integration/tcp_tunneling_integration_test.cc#L601-L604) is the test for this scenario). This is because with http1 upstream calling [encodeData](https://github.com/envoyproxy/envoy/blob/05e3529ebab0035310e8fd284ea5f05c7f3d27d8/source/common/http/http1/codec_impl.cc#L223)  on the encoder with `end_stream` set to `true` does not result in the connection with upstream being closed. For comparison, with http2 upstream calling [encodeData](https://github.com/envoyproxy/envoy/blob/05e3529ebab0035310e8fd284ea5f05c7f3d27d8/source/common/http/http2/codec_impl.cc#L476) with `end_stream` set to `true` results in a frame with `END STREAM` flag set.

After discussing this with @alyssawilk, we decided not to implement half-closed semantics for the moment and to reset the encoder on [doneWriting](https://github.com/envoyproxy/envoy/blob/05e3529ebab0035310e8fd284ea5f05c7f3d27d8/source/common/tcp_proxy/upstream.cc#L302). Unfortunately this has a the side-effect of closing the upstream connection without flushing ([TcpProxyUpstreamFlush](https://github.com/envoyproxy/envoy/blob/05e3529ebab0035310e8fd284ea5f05c7f3d27d8/test/integration/tcp_tunneling_integration_test.cc#L737-L740) integration test is revealing this issue).

@mattklein123 @snowp, do you have any suggestions on how to proceed?

mattklein123 commented:
> @mattklein123 @snowp, do you have any suggestions on how to proceed?

Sorry can you clarify what you are asking for? What are the current problems? I see the text above about half closed but I don't think I have enough context to help without more info.

irozzo-1A commented:
> Sorry can you clarify what you are asking for? What are the current problems? I see the text above about half closed but I don't think I have enough context to help without more info.

Sure. Hope this clarifies.

### Description of the scenario

When a tcp client connect with envoy acting as a TCP proxy with http1 tunneling config, a HTTP/1.1 CONNECT request is generated and sent to the upstream. If the upstream respond with a 2xx return code the connection between the envoy and the upstream proxy is hijacked for streaming the TCP data in both directions. When the downstream client closes its end of the socket (the write end), I expect Envoy to flush all the pending data and close its write end with the upstream proxy.

### What is happening

This is what happens when the downstream client finishes writing and closes it's end of the socket:

1. `Envoy::TcpProxy::Filter::onData(data, true)` the second argument is `end_stream`
2. `Envoy::TcpProxy::HttpUpstream::encodeData(data, true)`
3. `Envoy::Http::Http1::StreamEncoderImpl::encodeData(data, true)`

The problem is that this is not triggering the termination of the connection with the downstream.

[What I tried so far](https://github.com/envoyproxy/envoy/pull/13293/commits/e6978b409a435d4b0729b4179395452272a2a4fa), as I mentioned above, is to reset the stream when `encodeData` is called with `end_stream` set (see the backtrace below to better understand). Now, this fix tentative failed because the connection is closed with `NoFlush` as you can see in the backtrace below, and this may result in data loss.

What I'm asking is basically some guidance on how to approach this scenario as I'm not familiar yet with the codebase. e.g. naively I thought I could simply convey somehow the `FlushWrite` type to `Envoy::Network::ConnectionImpl::close`. I already discussed with @alyssawilk and I think she would probably suggest starting adding half-close semantics for HTTP/1.1, but she also suggested me to bring the topic here to see if you guys see a less complicated solution.

```
  * frame #0: 0x00000001049f51e2 tcp_tunneling_integration_test`Envoy::Network::ConnectionImpl::close(this=0x00000000038c5080, type=NoFlush) at connection_impl.cc:105:28
    frame #1: 0x00000001041482ea tcp_tunneling_integration_test`Envoy::Http::CodecClient::close(this=0x00000000038db050) at codec_client.cc:54:42
    frame #2: 0x00000001011020ea tcp_tunneling_integration_test`Envoy::Http::Http1::ConnPoolImpl::onDownstreamReset(this=0x00000000038f0360, client=0x0000000003829a80) at conn_pool.cc:46:25
    frame #3: 0x0000000101105588 tcp_tunneling_integration_test`Envoy::Http::Http1::ConnPoolImpl::StreamWrapper::onResetStream(this=0x000000000389b580, (null)=LocalReset, (null)=(__data = 0x0000000000000000, __size = 0)) at conn_pool.h:54:24
    frame #4: 0x000000010418928e tcp_tunneling_integration_test`Envoy::Http::StreamCallbackHelper::runResetCallbacks(this=0x00000000037debe8, reason=LocalReset) at codec_helper.h:49:20
    frame #5: 0x00000001041eb6fa tcp_tunneling_integration_test`Envoy::Http::Http1::ClientConnectionImpl::onResetStream(this=0x00000000037de880, reason=LocalReset) at codec_impl.cc:1296:40
    frame #6: 0x00000001041d4e37 tcp_tunneling_integration_test`Envoy::Http::Http1::ConnectionImpl::onResetStreamBase(this=0x00000000037de880, reason=LocalReset) at codec_impl.cc:835:3
    frame #7: 0x00000001041d4a6f tcp_tunneling_integration_test`Envoy::Http::Http1::StreamEncoderImpl::resetStream(this=0x00000000037debe0, reason=LocalReset) at codec_impl.cc:313:15
    frame #8: 0x0000000103fec279 tcp_tunneling_integration_test`Envoy::TcpProxy::HttpUpstream::resetEncoder(this=0x00000000037dad70, event=LocalClose, inform_downstream=true) at upstream.cc:115:35
    frame #9: 0x0000000103fef671 tcp_tunneling_integration_test`Envoy::TcpProxy::Http1Upstream::doneWriting(this=0x00000000037dad70) at upstream.cc:302:3
    frame #10: 0x0000000103fec40c tcp_tunneling_integration_test`Envoy::TcpProxy::HttpUpstream::encodeData(this=0x00000000037dad70, data=0x0000000003813740, end_stream=true) at upstream.cc:76:5
    frame #11: 0x0000000103fa7bdc tcp_tunneling_integration_test`Envoy::TcpProxy::Filter::onData(this=0x00000000038ea358, data=0x0000000003813740, end_stream=true) at tcp_proxy.cc:563:16
    frame #12: 0x0000000104a74e45 tcp_tunneling_integration_test`Envoy::Network::FilterManagerImpl::onContinueReading(this=0x00000000038136f8, filter=0x0000000000000000, buffer_source=0x0000000003813680) at filter_manager_impl.cc:66:48
    frame #13: 0x0000000104a75772 tcp_tunneling_integration_test`Envoy::Network::FilterManagerImpl::onRead(this=0x00000000038136f8) at filter_manager_impl.cc:76:3
    frame #14: 0x00000001049f7e11 tcp_tunneling_integration_test`Envoy::Network::ConnectionImpl::onRead(this=0x0000000003813680, read_buffer_size=0) at connection_impl.cc:292:19
    frame #15: 0x0000000104a017be tcp_tunneling_integration_test`Envoy::Network::ConnectionImpl::onReadReady(this=0x0000000003813680) at connection_impl.cc:574:5
    frame #16: 0x00000001049fdf1c tcp_tunneling_integration_test`Envoy::Network::ConnectionImpl::onFileEvent(this=0x0000000003813680, events=1) at connection_impl.cc:534:5
    frame #17: 0x0000000104a3ff0e tcp_tunneling_integration_test`Envoy::Network::ConnectionImpl::ConnectionImpl(this=0x00000000038ea2f8, events=1)::$_6::operator()(unsigned int) const at connection_impl.cc:71:54
```

mattklein123 commented:
> The problem is that this is not triggering the termination of the connection with the downstream.

Do you mean upstream?

Also, in the description above when you say "close" do you mean "half close?" Can you clarify in terms of full/half close semantics what you want the behavior to be being extremely careful about using downstream and upstream accurately?

alyssawilk commented:
Yeah, basically for HTTP/2 we can "end stream" by sending a FIN frame.  For HTTP/1 we want to signal the stream is ended but closing the connection, and there's no way to flush-close upstream.  My suggestion was that we add an allow-frame-by-connection-close to the HTTP/1.1 protocol options and if it's true, there's no content length or transfer encoding, the end stream bool would result in a Network:::Connection half close.

mattklein123 commented:
> My suggestion was that we add an allow-frame-by-connection-close to the HTTP/1.1 protocol options and if it's true, there's no content length or transfer encoding, the end stream bool would result in a Network:::Connection half close.

+1. It's possible we don't need a config option and can do something like how I dealt witht he hystrix disable chunk encoding setting? https://github.com/envoyproxy/envoy/blob/d043a1d01468592e1363be383347a4505ba90787/include/envoy/http/codec.h#L59

irozzo-1A commented:
> Do you mean upstream?

Yes, you're right I meant upstream. I fixed this in the original comment.

> Also, in the description above when you say "close" do you mean "half close?" Can you clarify in terms of full/half close semantics what you want the behavior to be being extremely careful about using downstream and upstream accurately?

To be as precise as I can, the diagram below represents what I expect to happen from an L4 standpoint. But what I got so far is either:

- No FIN sent from Envoy to Upstream (before https://github.com/envoyproxy/envoy/pull/13293/commits/e6978b409a435d4b0729b4179395452272a2a4fa)
- FIN sent but not all data transmitted because of no flush-close (after https://github.com/envoyproxy/envoy/pull/13293/commits/e6978b409a435d4b0729b4179395452272a2a4fa)

![envoy-http1-tunneling](https://user-images.githubusercontent.com/28569040/96708153-1b11af80-1399-11eb-9b6e-171ccd24fc48.png)

alyssawilk commented:
ah yeah, that'd work too.  Basically I think we want it "off by default" since we don't normally allow frame by close for requests and it'd limit the risk of the change.

mattklein123 commented:
I think there 2 different things being requested here though:
1) Possibly support half-close for HTTP/1.1 connections.
2) If we do a full-close, support flush write

I think (2) is easy and can be done the way @alyssawilk and I are suggesting (modulo some minor details). If we need to support (1) I think that will be a larger and more scary change.

Is (2) sufficient for now?

alyssawilk commented:
I think you have difficulty inverted.
I think for flush write it'd need to be flush write and delay, which means the lifetime of the upstream stream needs to be prolonged which is really hard.
For half close, I think doing it for this corner case ("if end stream and half close is enabled which it isn't by default and no content length and not chunked, FIN the underlying connection") is pretty straight-forward.  I think the only bit we need to do for upstream is make sure that if upstream closes the FIN doesn't arrive ahead of the data, which I think we already handle correctly.

mattklein123 commented:
> I think you have difficulty inverted.

Quite possible. :)

> I think for flush write it'd need to be flush write and delay, which means the lifetime of the upstream stream needs to be prolonged which is really hard.

True, though I wonder if we could get away with no delay initially? For supporting the delay, TCP proxy already supports a "flush write connection holder." Given that the connection pools now share a lot of code I wonder if we could just make this generic and then opt-in for this case?

> For half close, I think doing it for this corner case ("if end stream and half close is enabled which it isn't by default and no content length and not chunked, FIN the underlying connection") is pretty straight-forward. I think the only bit we need to do for upstream is make sure that if upstream closes the FIN doesn't arrive ahead of the data, which I think we already handle correctly.

Yeah, I just worry about the small details here and how many things might need to be fixed given that the HTTP code has never worked with half-close. It might just work. I'm guessing it won't. :)

alyssawilk commented:
yeah, I was looking at the TCP drain code but it takes raw connection pointer stuff and I think hooking that up would be a mess
The other question really is if we need to communicate half close.  We do need half close support for TCP in general so while I agree it's likely to result in full on close semantics at the HTTP/1/.1 connect-reading peer today, I think in the long run the same way that the client can half close TCP, we're going to need to support sending that half close through an HTTP CONNECT tunnel and decapsulating.  I think sending the FIN and acknowledging that it's only the first step on that journey and eventually for HTTP/1.1 connect we should take that FIN and convert it into a "stream ended" and half close the upstream TCP connection.  But that's a set of PRs for another day :-P

alyssawilk commented:
This ready for another pass?

irozzo-1A commented:
> This ready for another pass?

Yep, I have still to write docs, but codewise should be almost ready.

irozzo-1A commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -261,6 +263,11 @@ void StreamEncoderImpl::endEncode() {
   connection_.onEncodeComplete();
+  // If half-close semantics are enabled, no chunk encoding nor content length
+  // are given, we signal end stream by half-closing the connection.
+  if (connection_.connection().isHalfCloseEnabled() && !chunk_encoding_) {
```
I should add a check on `content_length`

alyssawilk replied:
yes please.

alyssawilk commented on source/common/http/codec_client.h:
```diff
@@ -173,8 +173,12 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
+    Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override {
       parent_.onData(data);
+      // Reset the codec client on end stream when HTTP1
+      if (end_stream && parent_.protocol() <= Protocol::Http11) {
```
One request, can we comment here that this will result in the connecting looking like it closed locally when it is functionally closed remotely?

I'd be mildly inclined to either put in a check to see if the connection has half close enabled here, or to add a test that in the normal case when the connection is closed we get the close event before an onData(... true) just to make sure the mis-attribution only applies for our CONNECT path.

irozzo-1A replied:
Ok, I'll start by adding the comment.

Regarding the check, I took a quick look but at first glance, I did not see an easy way of adding it. I will take another look.

irozzo-1A replied:
Well, adding a method `isHalfCloseEnabled` on the CodecClient would make it.

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -371,6 +378,7 @@ void RequestEncoderImpl::encodeHeaders(const RequestHeaderMap& headers, bool end
   const HeaderEntry* path = headers.Path();
   const HeaderEntry* host = headers.Host();
   bool is_connect = HeaderUtility::isConnect(headers);
+  setIsResponseToConnectRequest(is_connect);
```
In this case, isn't it the actual CONNECT request, not the response to CONNECT request?

irozzo-1A replied:
yeah, I was doing an experiment here because I thought that `content_length` and `chunk_encoding` headers should be ignored in this test:

https://github.com/envoyproxy/envoy/blob/da60c3726522138182226e9f0deaa8a1e17da4e8/source/common/http/http1/codec_impl.cc#L268

when the method is CONNECT as mentioned in the RFC, but I was not sure how to convey the method information. If you have any suggestions on how to handle this I'm taker, otherwise, I'll take another look.

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -222,7 +203,17 @@ bool HttpConnPool::valid() const { return conn_pool_ != nullptr; }
+    upstream_ = std::make_unique<Http2Upstream>(upstream_callbacks_, hostname_);
+    break;
+  case Http::CodecClient::Type::HTTP3:
+  default:
```
can we add validation in valid() that type is either HTTP1 or HTTP2?
Then here we can just say
if (http1) {
} else {
}
and not have uncovered crashy code.

alyssawilk replied:
ping?

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -248,5 +239,71 @@ void HttpConnPool::onPoolReady(Http::RequestEncoder& request_encoder,
+  // Any 'Content-Length' or 'Transfer-Encoding' header fields MUST be ignored.
+  // https://tools.ietf.org/html/rfc7231#section-4.3.6
+  const uint64_t rc = Http::Utility::getResponseStatus(headers);
+  if (rc < 200 || rc >= 300) {
```
CodeUtility::is2xx :-)

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -248,5 +239,71 @@ void HttpConnPool::onPoolReady(Http::RequestEncoder& request_encoder,
+    : HttpUpstream(callbacks, hostname), response_decoder_(*this) {}
+
+void Http1Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder,
+                                      bool /* isSSL unused here */) {
```
you can just have  bool)  without the comments

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -248,5 +239,71 @@ void HttpConnPool::onPoolReady(Http::RequestEncoder& request_encoder,
+  // According to RFC7231 any 2xx response indicates that the connection is
+  // established.
+  // Any 'Content-Length' or 'Transfer-Encoding' header fields MUST be ignored.
+  // https://tools.ietf.org/html/rfc7231#section-4.3.6
```
Also super bonus points for RFC links
I was about to comment that they're not allowed but you're right - they're not allowed but should be ignored :-)

Let's make sure we have an integration test for each just to make sure for example that the headers say "chunked" and the connect payload looks like http body, that it is passed through as CONNECT payload and not dechuked by HTTP parser.

irozzo-1A replied:
I tried to add an integration test where the upstream server replies with `TransferEncoding` header set, but it fails on this assertion:

https://github.com/envoyproxy/envoy/blob/d0ea3325dd96510ae785d2deef25e17059540672/source/common/http/http1/codec_impl_legacy.cc#L142

Maybe I can just try to add some unit test coverage here?

alyssawilk replied:
you'll probably have to use the raw connection code
see TestServerAllowChunkedLength for how to wait for request headers complete (\r\n\r\n) and send a raw HTTP/1.1 response.

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -248,5 +239,71 @@ void HttpConnPool::onPoolReady(Http::RequestEncoder& request_encoder,
+    ENVOY_LOG_MISC(trace, "Http1Upstream: invalid response. Return code: {}.", rc);
+    return false;
+  }
+  ENVOY_LOG_MISC(trace, "Http1Upstream: valid response. Return code: {}.", rc);
```
I'd be inclined to take these out but your call :-)

alyssawilk commented on source/common/tcp_proxy/upstream.h:
```diff
@@ -178,14 +181,39 @@ class HttpUpstream : public GenericUpstream, Http::StreamCallbacks {
+  Http::ResponseDecoder& responseDecoder() override { return response_decoder_; }
+
+private:
+  DecoderShim response_decoder_;
```
can this just be protected or private-with-accessors in the base class?

alyssawilk reviewed (commented):
This looks fantastic!!  Thanks so much for all the work you've put in here.
First pass on code today, and I'll do a round on tests when I'm back Monday :-)

irozzo-1A commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -459,18 +459,21 @@ bool Filter::maybeTunnel(const std::string& cluster_name) {
+      // TODO(irozzo) Currently we assume HTTP1 by the absence of HTTP2
+      // feature. This logic will break if unchanged when HTTP3 will be
+      // supported.
+      ENVOY_LOG(debug, "TCP:maybeTunnel(), using HTTP/1 upstream.");
```
Nope, forgot to remove it.

irozzo-1A commented:
@alyssawilk I think I addressed all comments so far, I just did not manage to add all integration tests required here https://github.com/envoyproxy/envoy/pull/13293#discussion_r515131921 and I'm not sure I did https://github.com/envoyproxy/envoy/pull/13293#discussion_r514365802 the right way (see https://github.com/envoyproxy/envoy/pull/13293/commits/39c4a7bd2ac603bccbf0a98a26ec1de8d2aae311), but I wait for your return. I added some docs too.

alyssawilk commented on docs/root/intro/arch_overview/http/upgrades.rst:
```diff
@@ -108,3 +108,18 @@ and `bazel-bin/source/exe/envoy-static --config-path  configs/terminate_connect.
+you will be running Envoy listening for TCP traffic on port 10000 and encapsulating it in an HTTP/1.1
+CONNECT request for an HTTP proxy running on localhost and listenig on port
+3128, having as a final destination dest.host.com on port 443.
+passing from .
```
yay docs!
is this an extra "passing from ."?

Alternately, do you think we should just have one section

Tunneling TCP over HTTP?

and take out the HTTP/2 specific comments above?

I think we could collapse the two sections, and just call out the two example configs.  WDYT?

irozzo-1A replied:
I'll give a try

alyssawilk commented on source/common/http/codec_client.h:
```diff
@@ -173,8 +179,15 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
+    Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override {
       parent_.onData(data);
+      // Reset the codec client on end stream when HTTP1.
+      if (end_stream && parent_.isHalfCloseEnabled() && parent_.protocol() <= Protocol::Http11) {
```
I'd be inclined to take out the protocol check here, since half close is only enabled for HTTP/1 (but if we added it for HTTP/2 we'd probably want to treat as full close)

irozzo-1A replied:
yes, I agree on this.

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -261,6 +264,11 @@ void StreamEncoderImpl::endEncode() {
   connection_.onEncodeComplete();
+  // If half-close semantics are enabled, no chunk encoding nor content length
+  // are given, we signal end stream by half-closing the connection.
+  if (connection_.connection().isHalfCloseEnabled() && !chunk_encoding_ && !content_length_) {
```
Do you think we should just replace this with
if (is_connect_) since connect requests will always need this?  I wonder if we can get rid of the new disable_content_length_ and the new content_length booleans

irozzo-1A replied:
yep, I like the idea of reducing the scope as this is the only use case at the moment.

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -136,6 +136,7 @@ void StreamEncoderImpl::encodeHeadersBase(const RequestOrResponseHeaderMap& head
 
   if (headers.ContentLength()) {
     saw_content_length = true;
+    content_length_ = !disable_content_length_;
```
so we're not really doing framing by half close here if we didn't see T-E chunked or content length, we're force doing it for connect requests.

I think we should either
1) just set half close for connect requests, and half-close for connect requests when we hit end stream
2) really do half close any time we don't have content length or te-chunked.

right now it looks like we're mixing and matching (or I'm misreading the code - please correct me if I'm wrong!)

irozzo-1A replied:
This was just to ignore the `content_length_` flag in the scope of a CONNECT request, not sure it was a good idea though :-)
I took the first approach you suggested.

alyssawilk commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -459,18 +459,21 @@ bool Filter::maybeTunnel(const std::string& cluster_name) {
-                                                        config_->tunnelingConfig()->hostname(),
-                                                        *upstream_callbacks_);
+    if ((cluster->info()->features() & Upstream::ClusterInfo::Features::HTTP2) != 0) {
+      // TODO(irozzo) Currently we assume HTTP1 by the absence of HTTP2
```
having not heard back let's just take this TODO out :-)

alyssawilk reviewed (commented):
Looking better and better!  here's a few other thoughts.

irozzo-1A commented:
@alyssawilk PTAL I think I addressed your comments

alyssawilk commented on docs/root/intro/arch_overview/http/upgrades.rst:
```diff
@@ -108,3 +111,14 @@ and `bazel-bin/source/exe/envoy-static --config-path  configs/terminate_connect.
+HTTP/1.1 CONNECT can be used to have TCP client connecting to its own
+destination passing through an HTTP proxy server (e.g. corporate proxy):
+
+[HTTP client] --- raw HTTP --- [Envoy]  --- HTTP tunneled over HTTP/1.1  --- [Upstream HTTP Proxy]  --- raw HTTP  --- [HTTP Server]
```
Let's have the client side and server side the same for both HTTP/1 and HTTP/2 (you can reverse either, but having upstream on the left farther above and client on the left immediately above seems confusing)

alyssawilk commented on source/common/http/codec_client.h:
```diff
@@ -63,6 +63,12 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
   }
 
+  /**
+   * Return flag indicating whether half-close semantics are enabled on the
```
Let's avoid "flag" here

Return if half-close semantics are enabled on the underlying connection
?

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -261,6 +261,10 @@ void StreamEncoderImpl::endEncode() {
 
   connection_.flushOutput(true);
   connection_.onEncodeComplete();
+  // With CONNECT half-closing the connection is used to signal end stream.
```
Mind 
With CONNECT -> With CONNECT,
just for readability?

alyssawilk commented on source/common/http/http1/codec_impl.h:
```diff
@@ -162,7 +163,6 @@ class RequestEncoderImpl : public StreamEncoderImpl, public RequestEncoder {
 private:
   bool upgrade_request_{};
   bool head_request_{};
-  bool connect_request_{};
```
why the rename?  we're only tracking CONNECT requests, not connect responses right?

irozzo-1A replied:
yep

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -54,17 +56,10 @@ TcpUpstream::onDownstreamEvent(Network::ConnectionEvent event) {
 HttpUpstream::HttpUpstream(Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
                            const std::string& hostname)
-    : upstream_callbacks_(callbacks), response_decoder_(*this), hostname_(hostname) {}
+    : hostname_(hostname), response_decoder_(*this), upstream_callbacks_(callbacks) {}
```
did the ordering need to change here?  If not let's revert.

irozzo-1A replied:
yep, this is needed because I moved `hostname_` to protected

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -222,7 +204,18 @@ bool HttpConnPool::valid() const { return conn_pool_ != nullptr; }
 void HttpConnPool::newStream(GenericConnectionPoolCallbacks* callbacks) {
   callbacks_ = callbacks;
-  upstream_ = std::make_unique<HttpUpstream>(upstream_callbacks_, hostname_);
+  switch (type_) {
```
Please make this 
if (http1) {
http1 upstream
} else {
http2 upstream
}
and check that type_ is valid when we create, and return false to valid() if type is not valid.

alyssawilk commented on test/integration/http_integration.h:
```diff
@@ -27,6 +27,8 @@ class IntegrationCodecClient : public Http::CodecClientProd {
                          Http::CodecClient::Type type);
 
   IntegrationStreamDecoderPtr makeHeaderOnlyRequest(const Http::RequestHeaderMap& headers);
```
I think you can just add an optional parameter to the original function but default it to false.

irozzo-1A replied:
good to know, I was unaware of this :-) I set the default to `true` to keep the behaviour unchanged

alyssawilk commented on test/integration/protocol_integration_test.cc:
```diff
@@ -1988,12 +1988,12 @@ TEST_P(DownstreamProtocolIntegrationTest, ConnectIsBlocked) {
   if (downstreamProtocol() == Http::CodecClient::Type::HTTP1) {
     // Because CONNECT requests for HTTP/1 do not include a path, they will fail
     // to find a route match and return a 404.
-    response->waitForEndStream();
```
why does this change?
If a connect request fails, it should still register as an end stream.

irozzo-1A replied:
you're right no need to change this

alyssawilk commented on test/integration/tcp_tunneling_integration_test.cc:
```diff
@@ -311,11 +311,12 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, ConnectTerminationIntegrationTest,
-class TcpTunnelingIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
-                                    public HttpIntegrationTest {
+// Tunneling downstream TCP over an upstream HTTP/2 CONNECT tunnel.
+class H2TcpTunnelingIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
```
Can we not just add a param of codec to this test?
I'd prefer we have all the same tests run for HTTP/1 and HTTP/2, modulo early return for any tests which have multiple streams for HTTP/1

I think we basically want all the same tests for HTTP/2 running for HTTP/1, and then we want a few "raw TCP" tests for HTTP/1 which test various content length and transfer encoding issues.

irozzo-1A replied:
Yep, I also thought that it's better to normalize it, I'll give it a try

alyssawilk reviewed (commented):
Looking awesome!  I think after this round I'm basically just down to test comments, so I'll tag a second reviewer once we have one more pass :-)

alyssawilk commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -459,18 +459,16 @@ bool Filter::maybeTunnel(const std::string& cluster_name) {
-                       "http2_protocol_options on the cluster.");
-      return false;
-    }
 
```
Huh, I just realized should be getting the HTTP connection pool from the cluster manager rather than creating one every time.  I'll go ahead and fix that once this lands, but can I ask you to write an HTTP/1 "no reuse" test (or tests) where we 

CONNECT, downstream close, new CONNECT request
and
CONNECT, upstream close, new CONNECT request

and make sure both the new CONNECT requests land on new upstream connections?  Kind of like the H2 connection reuse test but the reverse.

irozzo-1A replied:
Let me know if this is not what you meant: https://github.com/envoyproxy/envoy/pull/13293/commits/2e0857f5435ecd53c30d58c502fc5054808c3721

alyssawilk commented on test/integration/http_integration.h:
```diff
@@ -26,7 +26,8 @@ class IntegrationCodecClient : public Http::CodecClientProd {
                          Http::CodecClient::Type type);
 
-  IntegrationStreamDecoderPtr makeHeaderOnlyRequest(const Http::RequestHeaderMap& headers);
+  IntegrationStreamDecoderPtr makeHeaderOnlyRequest(const Http::RequestHeaderMap& headers,
```
header only request implies end stream is true.

I think you can change these calls to startRequest if you want headers without a FIN bit.

irozzo-1A replied:
Right, I did not see that method.

alyssawilk commented on test/integration/tcp_tunneling_integration_test.cc:
```diff
@@ -429,20 +429,31 @@ TEST_P(TcpTunnelingIntegrationTest, CloseUpstreamFirst) {
   // Send data from upstream to downstream with an end stream and make sure the data is received
   // before the connection is half-closed.
   upstream_request_->encodeData(12, true);
```
Do you think Envoy should do half close semantics when sending a CONNECT response?  If we did, the codec would frame with a half-close here and you wouldn't need the upstream close.

irozzo-1A replied:
yep, I think it could make sense, but I would tend to address it in a dedicated PR unless you consider this a blocker.

alyssawilk reviewed (commented):
Almost there.  Fantastic job with the tests, thank you!

alyssawilk commented:
Oh and please do a master merge since there's conflicts.

irozzo-1A commented:
> Almost there. Fantastic job with the tests, thank you!

Thx @alyssawilk! Unfortunately the integration test `TransferEncodingHeaderIgnoredHttp1` I introduced seems to be flaky, I'm checking.

irozzo-1A commented on test/integration/tcp_tunneling_integration_test.cc:
```diff
@@ -596,9 +633,175 @@ TEST_P(TcpTunnelingIntegrationTest, H2ConnectionReuse) {
+
+// TODO(irozzo): temporarily disabled as a protocol error is thrown when
+// transfer-encoding header is received in CONNECT responses.
+TEST_P(TcpTunnelingIntegrationTest, DISABLED_TransferEncodingHeaderIgnoredHttp1) {
```
I disabled this at the moment because of:

https://github.com/envoyproxy/envoy/blob/d5e8634edf26dab65218e42f2acfe7c3dc95b72d/source/common/http/http1/codec_impl.cc#L757

alyssawilk replied:
Yep, follow-up is fine.

irozzo-1A replied:
Personally, I'm not sure it's worth it. My plan was to try to address this point once this gets merged. WDYT @alyssawilk?

alyssawilk replied:
I think as long as we land an immediate follow-up (you could start working on it now!) it's fine by me.

irozzo-1A replied:
I would tend to give more priority to [this](https://github.com/envoyproxy/envoy/pull/13293#issuecomment-701408912), but yes I will try to take care of both ;-)

alyssawilk reviewed (dismissed):
LGTM with two meta-comments
1) can you update the PR description to be up to date with respect to unit tests etc.?
2) can you please avoid force-pushing in future?  Github doesn't handle force-push well.  If you merge rather than rebase you should be able to avoid it.

irozzo-1A commented:
> LGTM with two meta-comments

That's great, thx @alyssawilk !

> 1. can you update the PR description to be up to date with respect to unit tests etc.?

Done, I did not put release notes following your example here: #10623, because I thought that this cannot be considered prod until we address:

https://github.com/envoyproxy/envoy/pull/13293#issuecomment-701408912

One notable side-effect of not waiting for the CONNECT response (I noticed when manually testing with squid proxy as an upstream) is that the data sent right after the request could be interpreted as another request in case the response to the CONNECT request is not a 2xx, triggering a 400 from upstream that can make Envoy fail here:

https://github.com/envoyproxy/envoy/blob/7a15e5d200f9e4e5f5cfe0347fe1c337e534be4b/source/common/http/http1/codec_impl.h#L605

> 2. can you please avoid force-pushing in future?  Github doesn't handle force-push well.  If you merge rather than rebase you should be able to avoid it.

Ok, good to know. I will take this into account in the next PRs. I imagine I don't have to squash my commits then?

Just for my personal culture, what do you mean by

> Github doesn't handle force-push well

?

alyssawilk commented:
Indeed, not squshing anything that went upstream is also necessary to avoid force pushes, if you always just add commits (no sqashing, no merging) then the reviewer can look at diffs just from your recent pushes, and not have to look at the PR as a whole each time, which makes it easier to do incremental review on the bits that changed since last time.  Squashing local commits before you push to github is fine though :-)

antoniovicente commented on source/common/tcp_proxy/upstream.h:
```diff
@@ -44,10 +46,11 @@ class HttpConnPool : public GenericConnPool, public Http::ConnectionPool::Callba
   ~HttpConnPool() override;
 
-  bool valid() const { return conn_pool_ != nullptr; }
+  bool valid() const { return conn_pool_ != nullptr && type_ <= Http::CodecClient::Type::HTTP2; }
```
Worth a comment to explain why HTTP3 is not considered a valid upstream protocol?

What tests we have to cover the case of type_ set to HTTP3?

irozzo-1A replied:
well, HTTP3 upstreams are not supported yet AFAIK, maybe I should simply remove this test. WDYT?

alyssawilk replied:
I lean towards keeping the validation, since it future-proofs us from messed up configs and custom upstreams.

irozzo-1A replied:
ok, I will just add a comment saying that this is to make it future-proof then (unless you think I should add a unit-test but I don't see much value in it)

antoniovicente reviewed (commented):
Looks pretty good.  Just a few nits about some edge cases and tests.

irozzo-1A commented on source/extensions/upstreams/tcp/generic/config.cc:
```diff
@@ -19,16 +20,18 @@ TcpProxy::GenericConnPoolPtr GenericConnPoolFactory::createGenericConnPool(
+                                                        config.value(), upstream_callbacks,
+                                                        Http::CodecClient::Type::HTTP1);
+      }
+    }();
```
yep, totally agree on this

irozzo-1A commented on test/integration/tcp_tunneling_integration_test.cc:
```diff
@@ -596,9 +633,175 @@ TEST_P(TcpTunnelingIntegrationTest, H2ConnectionReuse) {
+}
+
+TEST_P(TcpTunnelingIntegrationTest, InvalidResponseHeadersHttp1) {
+  if (upstreamProtocol() == FakeHttpConnection::Type::HTTP1) {
```
You are right, this test does not make sense and overlaps with `InvalidResponseHeaders`, I will remove it.

irozzo-1A commented on test/integration/tcp_tunneling_integration_test.cc:
```diff
@@ -596,9 +633,175 @@ TEST_P(TcpTunnelingIntegrationTest, H2ConnectionReuse) {
+
+  // Send upgrade headers downstream, fully establishing the connection.
+  default_response_headers_.setStatus(enumToInt(Http::Code::IMUsed));
+  default_response_headers_.setContentLength(1024);
```
This test basically asserts that even if the `content-length` header is present in the CONNECT response, which semantically does not make much sense as the TCP connection is hijacked and HTTP semantics do not apply any longer, the response is considered valid (i.e. no stream reset).

I agree with your point that making the content length < than the data sent will make even more explicit that the header is ignored. Will do that.

irozzo-1A commented on configs/encapsulate_in_http1_connect.yaml:
```diff
@@ -0,0 +1,43 @@
+    - name: cluster_0
+      connect_timeout: 5s
+      http_protocol_options:
+        {}
```
> Are the previous 2 lines what enables use of HTTP1 as a tunneling protocol?
> 
yes they are 
> Worth referring to configs/encapsulate_in_connect.yaml in comments? Should that config be renamed to indicate that tunneling happens over http2.

ok, I will do 

> You may want to minimize other differences between these two example configs. For instance, use the same hostname and upstream port_value in the two configs.

same here

irozzo-1A commented:
> Looks pretty good. Just a few nits about some edge cases and tests.

Thx for your review @antoniovicente ! I think I addressed the points you raised, apart from:

https://github.com/envoyproxy/envoy/pull/13293#discussion_r520351145
https://github.com/envoyproxy/envoy/pull/13293#discussion_r520495872

where I would like also to have @alyssawilk opinion.

alyssawilk commented on configs/encapsulate_in_http2_connect.yaml:
```diff
@@ -26,10 +26,11 @@ static_resources:
           stat_prefix: tcp_stats
           cluster: "cluster_0"
           tunneling_config:
-            hostname: host.com
```
sorry, why 10002 when it's connecting to port 10001?

irozzo-1A replied:
My idea was to show that the destination port can be different from the port used by the upstream proxy, but maybe using `443` makes more sense to be consistent with `terminate_connect.yaml`

alyssawilk replied:
Sending  CONNECT foo.com:1234 to foo.com:1235 seems odd to me, but if you think it's worth explicitly testing for that how about a comment so other folks don't think it's just an off by one error :-)

irozzo-1A replied:
Actually, in this scenario, the CONNECT is sent to the upstream that is `127.0.0.1:10001` targeting `host.com:10002`. IMO it should be quite common to have the upstream proxy listening on a port and receiving CONNECT requests targeting a different one. What could be considered a bit odd maybe, is that the L2 proxy is receiving `CONNECT host.com:10002` and it is connecting instead to `www.google.com:443`.

alyssawilk reviewed (approved):
@snowp did you want to take a pass? if not I'll merge tomorrow.

snowp commented on configs/configgen.py:
```diff
@@ -139,5 +139,5 @@ def generate_config(template_path, template, output_file, **context):
 
 shutil.copy(os.path.join(SCRIPT_DIR, 'envoyproxy_io_proxy.yaml'), OUT_DIR)
-shutil.copy(os.path.join(SCRIPT_DIR, 'encapsulate_in_connect.yaml'), OUT_DIR)
+shutil.copy(os.path.join(SCRIPT_DIR, 'encapsulate_in_http2_connect.yaml'), OUT_DIR)
```
Should the new HTTP1 example be included here?

irozzo-1A replied:
not sure about this, WDYT @alyssawilk ?

alyssawilk replied:
Yeah, might as well for completeness.

irozzo-1A replied:
done

snowp commented on docs/root/intro/arch_overview/http/upgrades.rst:
```diff
@@ -94,17 +94,31 @@ will synthesize 200 response headers, and then forward the TCP data as the HTTP
 you will be running two Envoys, the first listening for TCP traffic on port 10000 and encapsulating it in an HTTP/2
 CONNECT request, and the second listening for HTTP/2 on 10001, stripping the CONNECT headers, and forwarding the
 original TCP upstream, in this case to google.com.
+
+HTTP/1.1 CONNECT can be used to have TCP client connecting to its own
+destination passing through an HTTP proxy server (e.g. corporate proxy):
+
+[HTTP Server] --- raw HTTP --- [Upstream HTTP Proxy]  --- HTTP tunneled over HTTP/1.1  --- [Envoy]  --- raw HTTP  --- [HTTP Client]
+
+Examples of such a set up can be found in the Envoy example config :repo:`directory <configs/>`
+If you run `bazel-bin/source/exe/envoy-static --config-path configs/encapsulate_in_http1_connect.yaml --base-id 1`
+you will be running Envoy listening for TCP traffic on port 10000 and encapsulating it in an HTTP/1.1
+CONNECT addressed to an HTTP proxy running on localhost and listenig on port
+10001, having as a final destination host.com on port 443.
```
As a first time reader this seems a bit confusing: which proxy are we configuring with the provided config? Both? Just the one labeled 'Envoy'?

irozzo-1A replied:
In this scenario, I made the assumption that the `Upstream HTTP Proxy` is not an Envoy instance, that's why I only labeled the proxy that is tunneling the traffic with `Envoy` and just provided one configuration `configs/encapsulate_in_http1_connect.yaml`.
The main driver of this PR is to support use cases where the upstream HTTP proxy is supporting HTTP/1 only e.g. #11308
Do you think I should make this more clear?

snowp replied:
Having re-read this a few times i realize my confusion stems from the fact that the flow is server -> client, not client -> server, and I was originally thinking of the encapsulation flow as originating from the client. 

I think I would still have this example make use of two Envoy instances (assuming we support this) and then mention that this can be used when the upstream is another proxy that does not support HTTP/2 (to justify why you would ever use HTTP/1.1). I think this would be clearer and would make the example complete, as its not relying on the reader knowing how to set up another local proxy.

irozzo-1A replied:
yep, this could make sense. If it's not a problem for you I could address this with a follow-up PR as I still have some work to do on this feature.

irozzo-1A replied:
I did this already as I had to do another change.

snowp reviewed (commented):
Looks good for the most part, just a couple comments

snowp commented:
Also this should probably have a release note

irozzo-1A commented:
> Also this should probably have a release note

I thought to add a release note in a follow-up PR after address two remaining issues. Find some context here: https://github.com/envoyproxy/envoy/pull/13293#issuecomment-724179168

irozzo-1A commented:
Thx for reviewing @snowp, I think I addressed your comments ;-)

irozzo-1A commented:
@alyssawilk @antoniovicente @snowp Could you please let me know if you have any points left?

snowp reviewed (approved):
LGTM, thanks!
